### ISSUE TITLE
Wrongly reading model stream answer resulting in a duplication of tool-calls with same tool id

### ISSUE THREAD

**Issue by MaximeNoulin** [vercel/ai#3953]
### Description

After some hours of investigations, I finally get to the bottom of this that drove me crazy.  
I'm thinking this bug could be a "source bug" of many other issues.

# Context

## Provider

I've used Scaleway Generative APIs  
https://www.scaleway.com/en/docs/ai-data/generative-apis/quickstart/
https://www.scaleway.com/en/generative-apis/  

## Models concerned

At least:
- meta/llama-3.1-70b-instruct:fp8
- meta/llama-3.1-8b-instruct:fp8

But other models could be impacted if they answer the same way.

I expect this problem to be non-provider related but model related.  
You should be able to reproduce it with any Llama 3.1 provider. 

# Possible other issues with same underlying bug

- https://github.com/vercel/ai/issues/3872

I think it may have other issues related but I haven't checked deeply.

# How to reproduce the bug?

Use the Weather Demo Chatbot: https://vercel.com/templates/next.js/nextjs-ai-chatbot  
With a provider providing a Llama 3.1 8B or 70B

# Description of the bug

In the UI : 2 weather widgets are displayed but one has crashed. We have a React error indicating that there is a duplicate key for 2 components.

<img width="1029" alt="image" src="https://github.com/user-attachments/assets/13389bd2-a0c5-4d78-8ad8-a94a54962863">

<img width="974" alt="image" src="https://github.com/user-attachments/assets/002302ae-100b-4a32-84d9-d36c7a17d6ae">

On the Provider side: we have an error `'{"object":"error","message":"This model only supports single tool-calls at once!","type":"BadRequestError","param":null,"code":400}'`

These are both resulting errors but **its not the source of the problem**.

To my investigation point, I think the bug concerns only the "stream features" like "streamText"

# Explanation of the bug

I've narrowed it to the first call to the provider.  
Call done:
```
curl -X POST "$PROVIDER/v1/chat/completions" \
-H "Authorization: Bearer $API_KEY" \
-H "Content-Type: application/json" \
-d '{"model":"llama-3.1-8b-instruct","parallel_tool_calls":false,"temperature":0,"messages":[{"role":"system","content":"You can use tools to anwser user request."},{"role":"user","content":"Give me the latest news on ai"}],"tools":[{"type":"function","function":{"name":"searchGoogle","description":"Search on the web using Google Search","parameters":{"type":"object","properties":{"query":{"type":"string","description":"the query to search"}},"required":["query"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}],"tool_choice":"auto","stream":true,"stream_options":{"include_usage":true}}'
```

The ReadableStream answered was:
```
data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"role":"assistant","content":""},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":226,"completion_tokens":0}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"id":"chatcmpl-tool-b3b307239370432d9910d4b79b4dbbaa","type":"function","index":0,"function":{"name":"searchGoogle"}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":233,"completion_tokens":7}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"{\"query\": \""}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":241,"completion_tokens":15}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"latest"}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":242,"completion_tokens":16}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":" news"}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":243,"completion_tokens":17}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":" on"}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":244,"completion_tokens":18}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":" ai\"}"}}]},"logprobs":null,"finish_reason":null}],"usage":{"prompt_tokens":226,"total_tokens":245,"completion_tokens":19}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":""}}]},"logprobs":null,"finish_reason":"tool_calls","stop_reason":128008}],"usage":{"prompt_tokens":226,"total_tokens":246,"completion_tokens":20}}

data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[],"usage":{"prompt_tokens":226,"total_tokens":246,"completion_tokens":20}}

data: [DONE]
```

This flow is equivalent to something like:
```
{
  "id": "chat-4a34e46ee37349d2a167b00a6abea435",
  "object": "chat.completion",
  "created": 1733154079,
  "model": "meta/llama-3.1-70b-instruct:fp8",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": null,
        "tool_calls": [
          {
            "id": "chatcmpl-tool-a0d9e21c09c5480f82d7a92da78ce3ca",
            "type": "function",
            "function": {
              "name": "searchGoogle",
              "arguments": "{\"query\": \"latest news on ai\"}"
            }
          }
        ]
      },
      "logprobs": null,
      "finish_reason": "tool_calls",
      "stop_reason": 128008
    }
  ],
  "usage": {
    "prompt_tokens": 226,
    "total_tokens": 247,
    "completion_tokens": 21
  },
  "prompt_logprobs": null
}
```

The bug appears because of this chunk:
```
data: {"id":"chat-2267f7e2910a4254bac0650ba74cfc1c","object":"chat.completion.chunk","created":1733162241,"model":"meta/llama-3.1-8b-instruct:fp8","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":""}}]},"logprobs":null,"finish_reason":"tool_calls","stop_reason":128008}],"usage":{"prompt_tokens":226,"total_tokens":246,"completion_tokens":20}}
```

The bug is in this function: https://github.com/vercel/ai/blob/main/packages/openai/src/openai-chat-language-model.ts#L359

While reading the stream in this function https://github.com/vercel/ai/blob/main/packages/openai/src/openai-chat-language-model.ts#L359, we get the following logs

## Caption of following logs

To ease debugging and explanation, I've created a log flow to describe the problem. Here is the caption to understand the log flow:
- `mappedToolCalls` defines the value of https://github.com/vercel/ai/blob/main/packages/openai/src/openai-chat-language-model.ts#L501
- `toolcall-1` is https://github.com/vercel/ai/blob/main/packages/openai/src/openai-chat-language-model.ts#L549
- `duplicate toolcall` is at https://github.com/vercel/ai/blob/main/packages/openai/src/openai-chat-language-model.ts#L605
- `add chunk in forwardstream` is at https://github.com/vercel/ai/blob/main/packages/ai/core/generate-text/run-tools-transformation.ts#L165

Log flow:
```
2024-12-02T18:16:29: stream step
2024-12-02T18:16:30: mappedToolCalls [
2024-12-02T18:16:30:   {
2024-12-02T18:16:30:     index: 0,
2024-12-02T18:16:30:     id: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:     type: 'function',
2024-12-02T18:16:30:     function: { name: 'searchGoogle' }
2024-12-02T18:16:30:   }
2024-12-02T18:16:30: ]
2024-12-02T18:16:30: toolcall-1 {
2024-12-02T18:16:30:   id: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   type: 'function',
2024-12-02T18:16:30:   function: { name: 'searchGoogle', arguments: '' }
2024-12-02T18:16:30: } {
2024-12-02T18:16:30:   index: 0,
2024-12-02T18:16:30:   id: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   type: 'function',
2024-12-02T18:16:30:   function: { name: 'searchGoogle' }
2024-12-02T18:16:30: }
2024-12-02T18:16:30: // check if tool call is complete ===> No
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: '{"query": "' } } ]
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: 'latest' } } ]
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: ' news' } } ]
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: ' on' } } ]
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: ' ai"}' } } ]
2024-12-02T18:16:30: duplicate toolcall {
2024-12-02T18:16:30:   id: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   type: 'function',
2024-12-02T18:16:30:   function: { name: 'searchGoogle', arguments: '{"query": "latest news on ai"}' }
2024-12-02T18:16:30: }
2024-12-02T18:16:30: add chunk in forwardstream {
2024-12-02T18:16:30:   type: 'tool-call',
2024-12-02T18:16:30:   toolCallType: 'function',
2024-12-02T18:16:30:   toolCallId: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   toolName: 'searchGoogle',
2024-12-02T18:16:30:   args: '{"query": "latest news on ai"}'
2024-12-02T18:16:30: }
2024-12-02T18:16:30: mappedToolCalls [ { index: 0, function: { arguments: '' } } ]
2024-12-02T18:16:30: duplicate toolcall {
2024-12-02T18:16:30:   id: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   type: 'function',
2024-12-02T18:16:30:   function: { name: 'searchGoogle', arguments: '{"query": "latest news on ai"}' }
2024-12-02T18:16:30: }
2024-12-02T18:16:30: add chunk in forwardstream {
2024-12-02T18:16:30:   type: 'tool-call',
2024-12-02T18:16:30:   toolCallType: 'function',
2024-12-02T18:16:30:   toolCallId: 'chatcmpl-tool-3d480d27469f4941808ae5f8db4c50fa',
2024-12-02T18:16:30:   toolName: 'searchGoogle',
2024-12-02T18:16:30:   args: '{"query": "latest news on ai"}'
2024-12-02T18:16:30: }
```

# Explanation of the log flow

The tool-call is streamed from the model but at a given point, the arguments became a valid parsableJSON and then the tool-call is send.
Just after that, another chunk is parsed with an empty argument: `mappedToolCalls [ { index: 0, function: { arguments: '' } } ]`

This results in the same tool-call be sent again. This then trigger a lot of different kind of bugs afterward.

# Ideas of resolution

- stop expecting that the tool-call is finished just because the args are parseable and find something more trustworthy (I think it would be the best solution)
- use a simple security so that a tool-call with the same id can't be send twice down the pipe

Don't hesitate to ask if anything is unclear. I'll be glad to help!

### AI provider

@ai-sdk/openai v1.0.5

### Additional context

Version of packages:
ai: v4.0.9
@ai-sdk/openai: v1.0.5

**Comment by MaximeNoulin** [comment]
I was thinking about it and as an idea of resolution, why can't we wait for the finish_reason property to be set to send the tool-call object?

This would means that the tool-call stream is finished and that would be better than the `isParsableJson(toolCall.function.arguments)`?

**Comment by lgrammel** [comment]
Thanks for the detailed bug report! This could indeed be the cause of several issues. The solution will be to be resilient to tool call parts after a tool call has been completed. I don't want to delay tool calls because this slows down the overall pipeline in latency-sensitive applications.

**Comment by MaximeNoulin** [comment]
@lgrammel you're right about not slowing down the overall pipeline but don't you be afraid that you could also have wrong execution of a tool-call with partial arguments?

For instance if the arguments became parseable, and even with correct required arguments, but missing some optional arguments because the model hasn't streamed it yet?

**Comment by lgrammel** [comment]
Once it's parseable, it's finished. If you add anything after a complete JSON, it breaks.

**Comment by MaximeNoulin** [comment]
@lgrammel are you sure?

If the stream chunk would look like:

```
mappedToolCalls [ { index: 0, function: { arguments: ' ai"' } } ]
```

It became parseable but the next chunk could be:

```
mappedToolCalls [ { index: 0, function: { arguments: ',"otherParam":"myValue"}' } } ]
```

Isn't it right?

**Comment by lgrammel** [comment]
The first won't parse bc the JSON is not complete.

**Comment by MaximeNoulin** [comment]
@lgrammel my bad you're absolutely right, sorry I'm not well awake ahah

**Comment by sailxjx** [comment]
I want to confirm if this issue has been fixed in the latest version of @ai-sdk/react 1.1.16?

I updated to this version but still encountered an error in streaming mode. The previous tool call matched the last incorrect tool result.

Since these logics are implemented in useChat, I cannot fix this issue in my own code.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
The report demonstrates a concrete failure in existing streaming tool-call handling: one provider tool call is incorrectly emitted twice. This is broken current behavior rather than a request for new functionality.

### Evidence
- A streamed Llama 3.1 response ends with an empty tool-call arguments chunk after the complete arguments have already been parsed.
- The OpenAI provider stream logic emits the same tool call ID twice, causing duplicate UI components, a React duplicate-key error, and a subsequent provider error.
- A maintainer explicitly acknowledged it as a bug and proposed making processing resilient to tool-call parts received after completion.
